src/Microsoft.AspNetCore.Mvc.Api.Analyzers/ActualApiResponseMetadataFactory.cs

```diff
@@ -0,0 +1,292 @@
+
+            foreach (var returnStatementSyntax in methodSyntax.DescendantNodes(_shouldDescendIntoChildren).OfType<ReturnStatementSyntax>())
+            {
+                if (returnStatementSyntax.IsMissing || returnStatementSyntax.Expression.IsMissing)
+                {
+                    // Ignore malformed return statements.
+                    continue;
+                }
+
+                var responseMetadata = InspectReturnStatementSyntax(
+                    symbolCache,
+                    semanticModel,
+                    returnStatementSyntax,
+                    cancellationToken);
+
+                if (responseMetadata != null)
+                {
+                    actualResponseMetadata.Add(responseMetadata.Value);
+                }
+                else
+                {
+                    allReturnStatementsReadable = false;
+                }
+            }
+
+            return allReturnStatementsReadable;
+        }
+
+        internal static ActualApiResponseMetadata? InspectReturnStatementSyntax(
+            in ApiControllerSymbolCache symbolCache,
+            SemanticModel semanticModel,
+            ReturnStatementSyntax returnStatementSyntax,
+            CancellationToken cancellationToken)
+        {
+            var returnExpression = returnStatementSyntax.Expression;
+            var typeInfo = semanticModel.GetTypeInfo(returnExpression, cancellationToken);
+            if (typeInfo.Type.TypeKind == TypeKind.Error)
+            {
+                return null;
+            }
+
+            var statementReturnType = typeInfo.Type;
+
+            if (!symbolCache.IActionResult.IsAssignableFrom(statementReturnType))
+            {
+                // Return expression is not an instance of IActionResult. Must be returning the "model".
+                return new ActualApiResponseMetadata(returnStatementSyntax, statementReturnType);
+            }
+
+            var defaultStatusCodeAttribute = statementReturnType
+                .GetAttributes(symbolCache.DefaultStatusCodeAttribute, inherit: true)
+                .FirstOrDefault();
+
+            var statusCode = GetDefaultStatusCode(defaultStatusCodeAttribute);
+            ITypeSymbol returnType = null;
+            switch (returnExpression)
+            {
+                case InvocationExpressionSyntax invocation:
+                    {
+                        // Covers the 'return StatusCode(200)' case.
+                        var result = InspectMethodArguments(symbolCache, semanticModel, invocation.Expression, invocation.ArgumentList, cancellationToken);
+                        statusCode = result.statusCode ?? statusCode;
+                        returnType = result.returnType;
+                        break;
+                    }
+
+                case ObjectCreationExpressionSyntax creation:
+                    {
+                        // Recover values from 'return new StatusCodeResult(200) case.
```

Comment:
nit: Recover is a wierd choice to use in comment here, 'recover' in a compiler/language context means something different.